### PR TITLE
Add Notion configurations and indexing Implementation

### PR DESCRIPTION
**Pull request by praveen-elastic**
## Part of https://github.com/elastic/connectors-py/issues/1958

This PR adds Indexing functionality to notion connector.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

#### Changes Requiring Extra Attention

- [ ] Security-related changes (encryption, TLS, SSRF, etc)
- [ ] New external service dependencies added.

### CODE REVIEW

File: connectors/sources/notion.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+            else:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def make_request(self, endpoint_key, method="GET", **kwargs):
+        """Execute a request to the Notion API"""
```
Comment (timgrein) [inline connectors/sources/notion.py:101]:
```suggestion
```

File: connectors/sources/notion.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+        self._logger = logger
+        self.notion_secret_key = self.configuration["notion_secret_key"]
+        self.retry_count = self.configuration["retry_count"]
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @cached_property
+    def _get_client(self):
+        """Authenticate with Notion API"""
```
Comment (timgrein) [inline connectors/sources/notion.py:52]:
```suggestion
```

File: connectors/sources/notion.py [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        return AsyncClient(
+            auth=self.notion_secret_key,
+            base_url=BASE_URL,
+        )
+
+    @cached_property
+    def session(self):
+        """Generate aiohttp client session with configuration fields.
```
Comment (timgrein) [inline connectors/sources/notion.py:60]: WDYM with `configuration fields`?

File: connectors/sources/notion.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        self.databases = self.configuration["databases"]
+        self._logger = logger
+        self._sleeps = CancellableSleeps()
+        self.concurrent_downloads = self.configuration["concurrent_downloads"]
+
+    def _set_internal_logger(self):
+        self.notion_client.set_logger(self._logger)
+
+    async def ping(self):
+        """Verify the connection with Notion"""
```
Comment (timgrein) [inline connectors/sources/notion.py:210]:
```suggestion
```

File: connectors/sources/notion.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        try:
+            await self.notion_client.fetch_owner()
+            self._logger.info("Successfully connected to Notion.")
+        except Exception:
+            self._logger.exception("Error while connecting to Notion.")
+            raise
+
+    async def close(self):
+        """Closes unclosed client"""
```
Comment (timgrein) [inline connectors/sources/notion.py:219]:
```suggestion
```

File: connectors/sources/notion.py [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+        await self.notion_client.close()
+
+    async def _remote_validation(self, entity_type, entity_titles):
```
Comment (timgrein) [inline connectors/sources/notion.py:222]: I think the name is a bit misleading here, you look for specific entity types and titles, right? Maybe it should be called something like `get_entities` or something similar?

File: connectors/sources/notion.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+                    "filter": {"value": "page", "property": "object"},
+                }
+            for database in self.databases:
+                yield {
+                    "query": "" if database == "*" else database,
+                    "filter": {"value": "database", "property": "object"},
+                }
+
+    async def get_docs(self, filtering=None):
+        """Executes the logic to fetch Notion objects in async manner.
```
Comment (timgrein) [inline connectors/sources/notion.py:426]: Could you maybe expand the docstring a little bit describing which objects are fetched?

**Review by timgrein — COMMENTED** [review]
Left some comments in a first pass

File: connectors/sources/notion.py (hunk 3 above)
Comment (praveen-elastic) [reply connectors/sources/notion.py]: Removed.

File: connectors/sources/notion.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+    @cached_property
+    def session(self):
+        """Generate aiohttp client session.
+
+        Returns:
+            aiohttp.ClientSession: An instance of Client Session
+        """
+        connector = aiohttp.TCPConnector(limit=MAX_CONCURRENT_CLIENT_SUPPORT)
+
+        return aiohttp.ClientSession(
```
Comment (seanstory) [inline connectors/sources/notion.py:66]: This doesn't seem to use any auth. Are the URLs we're using the Session for not secured?

File: connectors/sources/notion.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+        try:
+            path = ENDPOINTS.get(endpoint_key)
+            if path:
+                return await self._get_client.request(
+                    path=path, method=method, **kwargs
+                )
+        except APIResponseError as api_error:
+            self._logger.exception(f"Notion API Error - {api_error}")
+            raise
+        except aiohttp.ClientError as client_error:
```
Comment (seanstory) [inline connectors/sources/notion.py:109]: Should some 429/`Retry-After` handling be added here?

File: connectors/sources/notion.py [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def make_request(self, endpoint_key, method="GET", **kwargs):
+        try:
+            path = ENDPOINTS.get(endpoint_key)
+            if path:
+                return await self._get_client.request(
```
Comment (seanstory) [inline connectors/sources/notion.py:103]: I'd love to see Debug logging for what path and kwargs are being requested

File: connectors/sources/notion.py [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+                    "query": "" if page == "*" else page,
+                    "filter": {"value": "page", "property": "object"},
+                }
+            for database in self.databases:
+                yield {
+                    "query": "" if database == "*" else database,
+                    "filter": {"value": "database", "property": "object"},
+                }
+
+    async def get_docs(self, filtering=None):
```
Comment (seanstory) [inline connectors/sources/notion.py:419]: I'd like to see INFO logging either in `get_docs` or one level down to provide indication of how the sync is progressing. Like:
```
[INFO] - fetching Users
...
[INFO] - fetching Pages
...
[INFO] - fetching Databases
```
etc

**Review by seanstory — COMMENTED** [review]
Largely looks like a good start. Left some small comments.

File: connectors/sources/notion.py (hunk 8 above)
Comment (praveen-elastic) [reply connectors/sources/notion.py:66]: Yes, these are public URLs as stated in the [API docs](https://developers.notion.com/docs/working-with-files-and-media#retrieving-files-and-media-via-the-notion-api)

Comment (seanstory) [reply connectors/sources/notion.py:65]: perfect. Thanks for the explanation!

**Review by seanstory — APPROVED** [review]
Thanks! Looks like a good start to the connector

**Pull request merged**